httpclient/http_client.go

```diff
@@ -95,18 +107,46 @@ func (hc *HTTPClient) DoRequest(ctx context.Context, pld HTTPRequest) (*http.Res
 		}
 	}
 
-	resp, err := hc.Client.Do(req)
+	return hc.Client.Do(req)
+}
+
+func hasSuccessCode(resp *http.Response) bool {
+	return resp.StatusCode >= 200 && resp.StatusCode < 300
+}
+
+type RequestFuncion func(ctx context.Context, pld HTTPRequest) (*http.Response, error)
+type ParserFunction[T any] func(resp *http.Response) (T, error)
+type ErrorParserFunction func(statusCode int, body []byte) error
+
+func RequestModel[T any](
+	ctx context.Context,
+	req RequestFuncion,
+	pld HTTPRequest,
+	parser ParserFunction[T],
+	errorParser ErrorParserFunction,
+) (model T, err error) {
```

Comment:
When I saw this at first, I was sceptical. I think this is a little bit of an overkill to make such an abstraction, especially because the return type is steered by the parser function and is not known when looking at it at the first glance, which makes it really annoying.

But then I played with it a little bit and it turns out that IDE fills in the template type T with your parser return type and makes it much less annoying and I think it's really smart overall! If the team likes this approach, we could use it also in other libs where we have our own http client :)

Reply:
I'm glad that you find this function useful. To be clear, I did not introduce this just because I like to write complicated generic constructs which contain all types of brackets. 

The main purpose of this task was to ensure that `resp.Body` entities are closed properly, so I had to find out how to do it once, and avoid the situation when the next developer "forgets" to close it. 
- First of all, I didn't want to modify public interfaces and the `DoRequest` function type. 
- Then, I realized that I couldn't close the body before a model is parsed. 
- Moreover, I found out that errors (especially `arcErrors`) are parsed based on the response body. 

Summing these factors up I figured, that I have to write a function which contains all the three parts `DoRequest`, `modelParsing` and `errorParsing` in one function and use a defer to properly close the response body at the end. 
That's the reason why I decided to do it this way :)

Reply:
Got it, that's another reason why we should the same in our other libs :)